go-controller/pkg/util/external_gw_conntrack.go

```diff
@@ -0,0 +1,315 @@
+		}
+		return &iface, nil
+	}
+	return nil, errors.New("no usable interface found")
+}
+
+func readNDPMsg(msg ndp.Message) (mac net.HardwareAddr, err error) {
+	// Expect a neighbor advertisement message with a target link-layer
+	// address option.
+	na, ok := msg.(*ndp.NeighborAdvertisement)
+	if !ok {
+		err = fmt.Errorf("message is not a neighbor advertisement: %T", msg)
+		return
+	}
+	if len(na.Options) != 1 {
+		err = fmt.Errorf("expected one option in neighbor advertisement")
+		return
+	}
+	lla, ok := na.Options[0].(*ndp.LinkLayerAddress)
+	if !ok {
+		err = fmt.Errorf("option is not a link-layer address: %T", msg)
+		return
+	}
+	return lla.Addr, nil
+}
+
+// getIPv6MacOnIface tries to resolve as many ips as possible.
+// Errors that prevent only 1 ip from being resolved are logged and not returned.
+// When an error is returned, some MACs may still be resolved and returned too.
+func getIPv6MacOnIface(info *ifaceWithTargetIPs) ([]net.HardwareAddr, error) {
+	// Set up an *ndp.Conn, bound to this interface's link-local IPv6 address.
+	c, _, err := ndp.Listen(info.iface, ndp.LinkLocal)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial NDP connection: %v", err)
+	}
+	defer c.Close()
+	for _, resolveIP := range info.ips {
+		target, err := netip.ParseAddr(resolveIP)
+		if err != nil {
+			klog.Errorf("Failed to ParseAddr %v: %v", resolveIP, err)
+			continue
+		}
+		// Use target's solicited-node multicast address to request that the target
+		// respond with a neighbor advertisement.
+		snm, err := ndp.SolicitedNodeMulticast(target)
+		if err != nil {
+			klog.Errorf("Failed to determine solicited-node multicast address: %v", err)
+			continue
+		}
+
+		// Build a neighbor solicitation message, indicate the target's link-local
+		// address, and also specify our source link-layer address.
+		m := &ndp.NeighborSolicitation{
+			TargetAddress: target,
+			Options: []ndp.Option{
+				&ndp.LinkLayerAddress{
+					Direction: ndp.Source,
+					Addr:      info.iface.HardwareAddr,
+				},
+			},
+		}
+
+		// Send the multicast message and wait for a response.
+		if err = c.WriteTo(m, nil, snm); err != nil {
+			klog.Errorf("Failed to write neighbor solicitation: %v", err)
+			continue
+		}
+	}
+	ipsToFind := sets.New[string](info.ips...)
+	macs := []net.HardwareAddr{}
+
+	// set max attempts to msgTries to limit the number of messages waiting for expected reply
+	for i := 0; i < msgTries*len(info.ips); i++ {
+		if err = c.SetReadDeadline(time.Now().Add(msgTimeout)); err != nil {
+			return macs, fmt.Errorf("failed to set read deadline: %w", err)
+		}
+		msg, _, ip, err := c.ReadFrom()
+		if err != nil {
+			// if some ips are resolved and others are not available anymore, return macs
+			// when no more messages are received
+			return macs, fmt.Errorf("failed to read NDP message: %v", err)
+		}
+		// target ip doesn't have a zone set, return ip without a zone to compare
+		ip = ip.WithZone("")
+
+		if ipsToFind.Has(ip.String()) {
```

Comment:
I'm wondering if this is not totally correct. The ip returned from ReadFrom is the sender IP, then you are comparing that with the target IPs you are interested in. However, I think the sender IP could be a link local IP, and the resolved target  address in the advertisement is the actual IP you care about.